core/src/Streamly/Internal/FileSystem/Posix/ReadDir.hsc

```diff
@@ -138,41 +257,64 @@ isMetaDir dname = do
                 then return True
                 else return False
 
-{-# NOINLINE lstatDname #-}
-lstatDname :: PosixPath -> Ptr CChar -> IO (Bool, Bool)
-lstatDname parent dname = do
+data GStatRes
+    = GSIsMetaDir
+    | GSIsRegDir
+    | GSIsNotDir
+    | GSIgnoreError Errno
+
+{-# NOINLINE gstatDname #-}
+gstatDname
+    :: ReadOptions -> PosixPath -> Ptr CChar -> IO GStatRes
+gstatDname conf parent dname = do
     isMeta <- liftIO $ isMetaDir dname
     if isMeta
-    then pure (True, True)
+    then pure GSIsMetaDir
     else do
         -- XXX We can create a pinned array right here since the next call pins
         -- it anyway.
         path <- appendCString parent dname
         Array.asCStringUnsafe (Path.toChunk path) $ \cStr -> do
-            res <- c_lstat_is_directory cStr
+            res <-
+                if _followSymlinks conf
+                then statWith c_stat cStr
+                else statWith c_lstat cStr
```

Comment:
you can just pass the flag to statWith and decide to use c_stat or c_lstat inside it. that will generate slightly better code.